src/nvme/nbft.c

```diff
@@ -0,0 +1,947 @@
+	}
+	get_heap_obj(raw_hfi_info_tcp, host_name, 1, &hfi->tcp_info.host_name);
+	if (raw_hfi_info_tcp->transport_flags & HFIINFOTCPFLAG_GLOBAL_ROUTE)
+		hfi->tcp_info.this_hfi_is_default_route = true;
+	return 0;
+}
+
+static int read_hfi(struct nbft_info *nbft, struct raw_nbft_hfi *raw_hfi, struct nbft_hfi **h)
+{
+	int ret;
+	struct nbft_hfi *hfi;
+	struct raw_nbft_header *header = (struct raw_nbft_header *)nbft->raw_nbft;
+
+	if (!(raw_hfi->hfi_flags & HFIFLAG_VALID))
+		return -EINVAL;
+	verify(raw_hfi->structure_id == NBFT_HFI, "invalid ID in HFI descriptor");
+
+	hfi = calloc(1, sizeof(struct nbft_hfi));
+	if (!hfi) {
+		return -ENOMEM;
+	}
+
+	hfi->index = raw_hfi->index;
+
+	/*
+	 * read HFI transport descriptor for this HFI
+	 */
+	if (raw_hfi->hfi_transport_type == nbft_trtype_tcp) {
+		/*
+		 * tcp
+		 */
+		struct raw_nbft_hfi_info_tcp *raw_hfi_info_tcp;
+
+		strncpy(hfi->transport, trtype_to_string(raw_hfi->hfi_transport_type), sizeof(hfi->transport));
+
+		ret = get_heap_obj(raw_hfi, hfi_transport_descriptor, 0, (char **)&raw_hfi_info_tcp);
+		if (ret)
+			goto fail;
+
+		ret = read_hfi_info_tcp(nbft, raw_hfi_info_tcp, hfi);
+		if (ret)
+			goto fail;
+	} else {
+		nvme_msg(NULL, LOG_DEBUG, "file %s: invalid transport type %d\n", nbft->filename, raw_hfi->hfi_transport_type);
+		ret = -EINVAL;
+		goto fail;
+	}
+
+	*h = hfi;
+	return 0;
+
+fail:
+	free(hfi);
+	return ret;
+}
+
+static int read_discovery(struct nbft_info *nbft, struct raw_nbft_discovery *raw_discovery, struct nbft_discovery **d)
+{
+	int ret;
+	struct nbft_discovery *discovery;
+	struct raw_nbft_header *header = (struct raw_nbft_header *)nbft->raw_nbft;
+
+	if (!(raw_discovery->flags & DISCOVERYFLAG_VALID))
+		return -EINVAL;
+	verify(raw_discovery->structure_id == NBFT_DISCOVERY, "invalid ID in discovery descriptor");
+
+	discovery = calloc(1, sizeof(struct nbft_discovery));
+	if (!discovery) {
+		ret = -ENOMEM;
+		goto discovery_fail;
+	}
+
+	discovery->index = raw_discovery->index;
+
+	if (get_heap_obj(raw_discovery, discovery_ctrl_addr, 1, &discovery->uri))
+		return -EINVAL;
+
+	if (get_heap_obj(raw_discovery, discovery_controller_nqn, 1, &discovery->nqn))
+		return -EINVAL;
+
+	discovery->hfi = hfi_from_index(nbft, raw_discovery->hfi_index);
+	if (raw_discovery->hfi_index && !discovery->hfi)
+		nvme_msg(NULL, LOG_DEBUG, "file %s: discovery %d HFI not found\n",
+			nbft->filename, discovery->index);
+
+	discovery->security = security_from_index(nbft, raw_discovery->security_index);
+	if (raw_discovery->security_index && !discovery->security)
+		nvme_msg(NULL, LOG_DEBUG, "file %s: discovery %d security descriptor not found\n",
+			nbft->filename, discovery->index);
+
+	*d = discovery;
+	return 0;
+
+discovery_fail:
+	free(discovery);
+	return ret;
+}
+
+static int read_security(struct nbft_info *nbft, struct raw_nbft_security *raw_security, struct nbft_security **s)
+{
+	/*
+	 *  TO DO add security stuff
+	 */
```

Comment:
Please either add the code or just leave those empty functions away.

Reply:
Looking at `struct nbft_security` there's substantial number of elements and objects to be parsed and translated. Better be addressed in a separate PR.

Reply:
NVMe security support is not are part of NVMe Boot 1.0.  There is more work that needs to be done in specification and in the EDK2 code to support NVMe security. I think we can ignore this stuff for now.  I will remove the TO DO.

Reply:
This To Do comment still remains.